static/js/components/dashboard/courses/StatusMessages.js

```diff
@@ -0,0 +1,155 @@
+// @flow
+import React from 'react';
+import R from 'ramda';
+import moment from 'moment';
+
+import type { CourseRun } from '../../../flow/programTypes';
+import {
+  STATUS_PASSED,
+  STATUS_NOT_PASSED,
+  STATUS_CAN_UPGRADE,
+  STATUS_MISSED_DEADLINE,
+  STATUS_CURRENTLY_ENROLLED,
+  STATUS_OFFERED,
+  STATUS_WILL_ATTEND,
+  STATUS_PENDING_ENROLLMENT,
+  DASHBOARD_FORMAT,
+  EDX_LINK_BASE,
+  COURSE_ACTION_PAY,
+  COURSE_ACTION_ENROLL,
+  COURSE_ACTION_REENROLL,
+} from '../../../constants';
+import { S } from '../../../lib/sanctuary';
+import {
+  courseCurrentlyInProgress,
+  hasPaidForAnyCourseRun,
+  courseUpcomingOrCurrent,
+  hasPassedCourseRun,
+  hasFailedCourseRun,
+} from './util';
+import { hasPassingExamGrade } from '../../../lib/grades';
+
+export const formatAction = message => (
+  <div className="second-col">
+    { message.action }
+  </div>
+);
+
+export const formatMessage = (message, index) => (
+  <div className="status-message cols" key={index}>
+    <div className="message first-col">
+      { message.message }
+    </div>
+    { message.action ? formatAction(message) : null }
+  </div>
+);
+
+// this calculates any status messages we'll need to show the user
+// we wrap the array of messages in a Maybe, so that we can indicate
+// the case where there are no messages cleanly
+// 
+// I'm so sorry for anyone who has to read / modify this
+// sometimes there really isn't a better way :/
+const calculateMessages = props => {
+  const {
+    courseAction,
+    firstRun,
+    course,
+    expandedStatuses,
+    setShowExpandedCourseStatus,
+  } = props;
```

Comment:
the overall component here is implemented as a functional pipeline and this is the first function in the pipeline, so if we wanted to refactor this to use separate arguments we would have to put some sort of wrapper around it to basically pull things out of `props` and pass them in to the calculate messages function, something like:

```js
calculateWrapper = props => {
  const  {
    courseAction,
    firstRun,
    course,
    expandedStatuses,
    setShowExpandedCourseStatus,
  } = props;
  
  return calculateMessages(
    courseAction,
    firstRun,
    course,
    expandedStatuses,
    setShowExpandedCourseStatus,
  );
}
```

which seems kind of overly boilerplate-ey to me. what do you think?